File: fidouafclient/src/main/java/io/hanko/fidouafclient/client/op/Authentication.java
package io.hanko.fidouafclient.client.op;

import android.content.Context;
import android.util.Base64;
import android.util.Log;

import com.google.gson.Gson;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import io.hanko.fidouafclient.asm.msgs.Request;
import io.hanko.fidouafclient.asm.msgs.request.ASMRequestAuth;
import io.hanko.fidouafclient.asm.msgs.request.AuthenticateIn;
import io.hanko.fidouafclient.asm.msgs.response.AuthenticateOut;
import io.hanko.fidouafclient.client.MainActivity;
import io.hanko.fidouafclient.client.interfaces.AsmStart;
import io.hanko.fidouafclient.client.interfaces.FacetIds;
import io.hanko.fidouafclient.client.msg.AuthenticationResponse;
import io.hanko.fidouafclient.client.msg.AuthenticatorSignAssertion;
import io.hanko.fidouafclient.client.msg.ChannelBinding;
import io.hanko.fidouafclient.client.msg.FinalChallengeParams;
import io.hanko.fidouafclient.client.msg.MatchCriteria;
import io.hanko.fidouafclient.client.msg.UafAuthenticationRequest;
import io.hanko.fidouafclient.client.msg.Version;
import io.hanko.fidouafclient.client.msg.client.UAFIntentType;
import io.hanko.fidouafclient.utility.ErrorCode;
import io.hanko.fidouafclient.utility.FidoUafUtils;
import io.hanko.fidouafclient.utility.GetAsmResponse;

public class Authentication implements FacetIds {

    private String TAG = "Authentication";
    private Gson gson;
    private AsmStart activity;
    private String facetId;
    private String channelBinding;
    private UafAuthenticationRequest authenticationRequest;
    private String appID;
    private FinalChallengeParams finalChallengeParams;
    private Context mContext;

    public Authentication(Context context, final AsmStart activity, final String facetId, final String channelBinding) {
        this.activity = activity;
        this.mContext = context;
        this.facetId = facetId;
        this.channelBinding = channelBinding;
        this.gson = new Gson();
    }

    public void processRequests(final UafAuthenticationRequest[] authenticationRequests) {
        for (UafAuthenticationRequest authReq : authenticationRequests) {
            if (authReq.getHeader().getUpv().getMajor() == 1 && authReq.getHeader().getUpv().getMinor() == 0) {
                authenticationRequest = authReq;
                break;
            }
            if (authReq.getHeader().getUpv().getMajor() == 1 && authReq.getHeader().getUpv().getMinor() == 0) {
                authenticationRequest = authReq;
            }
        }

        if (authenticationRequest == null) {
            activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.PROTOCOL_ERROR, null);
        } else if (authenticationRequest.getHeader().getAppID() == null || authenticationRequest.getHeader().getAppID().isEmpty() || Objects.equals(authenticationRequest.getHeader().getAppID(), facetId)) {
            appID = facetId;
            sendToAsm(true);
        } else if (authenticationRequest.getHeader().getAppID().contains("https://")){
            appID = authenticationRequest.getHeader().getAppID();
            FidoUafUtils.GetTrustedFacetsTask getTrustedFacetsTask = new FidoUafUtils.GetTrustedFacetsTask(this);
            getTrustedFacetsTask.execute(authenticationRequest.getHeader().getAppID());
        } else {
            activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.PROTOCOL_ERROR, null);
        }
    }
    
    public void processASMResponse(final AuthenticateOut authenticateOut) {
        AuthenticatorSignAssertion authenticationAssertion = new AuthenticatorSignAssertion(authenticateOut.assertion, authenticateOut.assertionScheme, null);
//        authenticationAssertion.assertion = authenticateOut.assertion;
//        authenticationAssertion.assertionScheme = authenticateOut.assertionScheme;

        AuthenticationResponse[] authenticationResponse = new AuthenticationResponse[1];
        authenticationResponse[0] = new AuthenticationResponse(
                authenticationRequest.getHeader(),
                Base64.encodeToString(gson.toJson(finalChallengeParams).getBytes(StandardCharsets.UTF_8), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING),
                Collections.singletonList(authenticationAssertion)
        );
//        authenticationResponse[0].header = authenticationRequest.getHeader();
//        authenticationResponse[0].fcParams = Base64.encodeToString(gson.toJson(finalChallengeParams).getBytes(StandardCharsets.UTF_8), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
//        authenticationResponse[0].assertions = new AuthenticatorSignAssertion[]{authenticationAssertion};

        activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.NO_ERROR, gson.toJson(authenticationResponse));
    }

    private void sendToAsm(final boolean isFacetIdValid) {
        if (isFacetIdValid) {
            try { // try-catch as workaround for Huawei smartphones, because they won´t call AsmStart.sendToAsm
                List<String> keyIds = new ArrayList<>();

                for (List<MatchCriteria> matchCriterias : authenticationRequest.getPolicy().getAccepted()) {
                    for (MatchCriteria matchCriteria : matchCriterias) {
                        if (matchCriteria.getKeyIDs() != null && matchCriteria.getKeyIDs().size() > 0) {
                            keyIds.addAll(matchCriteria.getKeyIDs());
                        }
                    }
                }

                finalChallengeParams = new FinalChallengeParams(appID, authenticationRequest.getChallenge(), facetId, gson.fromJson(channelBinding, ChannelBinding.class));
//                finalChallengeParams.appID = appID;
//                finalChallengeParams.challenge = authenticationRequest.getChallenge();
//                finalChallengeParams.facetID = facetId;
//                finalChallengeParams.channelBinding = gson.fromJson(channelBinding, ChannelBinding.class);

                GetAsmResponse response = FidoUafUtils.getAsmFromKeyId(mContext, appID, keyIds.toArray(new String[keyIds.size()]));

                AuthenticateIn authenticateIn = new AuthenticateIn();
                authenticateIn.appID = appID;
                authenticateIn.finalChallenge = Base64.encodeToString(gson.toJson(finalChallengeParams).getBytes(StandardCharsets.UTF_8), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
                if (response == null) {
                    activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.NO_SUITABLE_AUTHENTICATOR, null);
                    return;
                }
                if (response.keyId != null) {
                    authenticateIn.keyIDs = new String[]{response.keyId};
                }
                if (authenticationRequest.getTransaction() != null && authenticationRequest.getTransaction().size() > 0) {
                    authenticateIn.transaction = authenticationRequest.getTransaction().get(0);
                } else {
                    authenticateIn.transaction = null;
                }

                ASMRequestAuth asmRequestAuth = new ASMRequestAuth();
                asmRequestAuth.asmVersion = authenticationRequest.getHeader().getUpv();
                asmRequestAuth.requestType = Request.Authenticate;
                asmRequestAuth.authenticatorIndex = 0;
                asmRequestAuth.args = authenticateIn;

                activity.sendToAsm(gson.toJson(asmRequestAuth), MainActivity.ASM_AUTH_REQUEST_CODE, response.asmClass);
            } catch (Exception e) {
                Log.e(TAG, "Error while creating asmRequest or sending to ASM", e);
                activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.NO_SUITABLE_AUTHENTICATOR, null);
            }
        } else {
            activity.sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.UNTRUSTED_FACET_ID, null);
        }
    }

    @Override
    public void processTrustedFacetIds(final String trustedFacetJson) {
        sendToAsm(FidoUafUtils.isFacetIdValid(trustedFacetJson, new Version(1, 0), facetId));
    }
}

File: fidouafclient/src/main/java/io/hanko/fidouafclient/client/op/Authentication.kt
package io.hanko.fidouafclient.client.op

import android.util.Base64
import io.hanko.fidouafclient.asm.msgs.RequestType
import io.hanko.fidouafclient.asm.msgs.request.ASMRequestAuth
import io.hanko.fidouafclient.asm.msgs.request.AuthenticateIn
import io.hanko.fidouafclient.asm.msgs.response.AuthenticateOut
import io.hanko.fidouafclient.authenticator.config.AuthenticatorMetadata
import io.hanko.fidouafclient.client.msg.*
import io.hanko.fidouafclient.client.msg.client.ErrorCode
import io.hanko.fidouafclient.client.msg.client.UAFIntentType
import io.hanko.fidouafclient.util.ClientUtil
import io.hanko.fidouafclient.util.Crypto
import io.hanko.fidouafclient.util.Util
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.nio.charset.StandardCharsets

class Authentication(val facetId: String, val channelBinding: String) {

    private val TAG = "Authentication"
    private var authenticationRequest: UafAuthenticationRequest? = null
    private var appID: String? = null
    private var finalChallengeParams: FinalChallengeParams? = null
    private val mainScope = CoroutineScope(Dispatchers.Main + Job())

    fun processRequests(authenticationRequests: List<UafAuthenticationRequest>, sendToAsm: (String) -> Unit, sendReturnIntent: (UAFIntentType?, ErrorCode, String?) -> Unit, skipTrustedFacetValidation: Boolean) {
        val authReq = authenticationRequests.find { it.header.upv.major == 1 && it.header.upv.minor == 1 }
                ?: authenticationRequests.find { it.header.upv.major == 1 && it.header.upv.minor == 0 }
        authenticationRequest = authReq

        val challengeBytes = Base64.decode(authReq?.challenge ?: "", Base64.URL_SAFE)

        if (authReq == null) {
            sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.UNSUPPORTED_VERSION, null)
        } else if (authReq.policy.accepted.any { it.any { !it.isValid() } } ||
                authReq.policy.disallowed?.any { !it.isValid() } == true ||
                authReq.policy.accepted.isEmpty() ||
                challengeBytes.size < 8 || challengeBytes.size > 64 ||
                !validateTransaction(authReq.transaction)) {
            sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.PROTOCOL_ERROR, null)
        } else if (authReq.header.appID == null || authReq.header.appID.isEmpty() || authReq.header.appID == facetId) {
            appID = facetId
            this.sendToAsm(authReq, sendToAsm, sendReturnIntent)
        } else if (Util.isValidHttpsUrl(authReq.header.appID)) {
            appID = authReq.header.appID
            if (skipTrustedFacetValidation) {
                this.sendToAsm(authReq, sendToAsm, sendReturnIntent)
            } else {
                mainScope.launch {
                    val trustedFacets = ClientUtil.getTrustedFacetsAsync(appID!!)
                    if (trustedFacets != null && ClientUtil.isFacetIdValid(trustedFacets, Version(1, 0), facetId))
                        this@Authentication.sendToAsm(authReq, sendToAsm, sendReturnIntent)
                    else
                        sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.UNTRUSTED_FACET_ID, null)
                }
            }
        } else {
            sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.PROTOCOL_ERROR, null)
        }
    }

    private fun validateTransaction(transaction: List<Transaction>?): Boolean {
        if (transaction == null) {
            return true
        } else if (transaction.none { it.contentType == AuthenticatorMetadata.authenticator.tcDisplayContentType } ||
                transaction.filter { it.contentType == AuthenticatorMetadata.authenticator.tcDisplayContentType }.any { it.content.length > 200 } ||
                transaction.any { !isTransactionContentValid(it.content) }) {
            return false
        }

        return true
    }

    private fun isTransactionContentValid(content: String): Boolean {
        return try {
            Base64.decode(content, Base64.URL_SAFE).isNotEmpty() && !content.contains("/") && !content.contains("+") && !content.contains("=")
        } catch (ex: Exception) {
            false
        }
    }

    fun processASMResponse(authenticateOut: AuthenticateOut, sendReturnIntent: (UAFIntentType?, ErrorCode, String?) -> Unit) {
        val authenticationAssertion = AuthenticatorSignAssertion(authenticateOut.assertionScheme, authenticateOut.assertion, null)

        val authenticationResponse = listOf(AuthenticationResponse(
                authenticationRequest!!.header,
                Base64.encodeToString(Util.moshi.adapter(FinalChallengeParams::class.java).toJson(finalChallengeParams).toByteArray(StandardCharsets.UTF_8), Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING),
                listOf(authenticationAssertion)
        ))

        sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.NO_ERROR, Util.moshi.adapter(Array<AuthenticationResponse>::class.java).toJson(authenticationResponse.toTypedArray()))
    }

    private fun sendToAsm(authenticationRequest: UafAuthenticationRequest, sendToAsm: (String) -> Unit, sendReturnIntent: (UAFIntentType?, ErrorCode, String?) -> Unit) {
        val allowedMatchCriterias = authenticationRequest.policy.accepted.filter {
            return@filter if (it.size == 1) // check that secondary list has one element, because we only support use of the embedded authenticator
                it.any { matchCriteria -> matchCriteria.matchesAuthenticator(AuthenticatorMetadata.authenticator, appID!!, false) }
            else
                false
        }.flatten()
        val disallowedMatchCriterias = authenticationRequest.policy.disallowed?.filter { it.matchesAuthenticator(AuthenticatorMetadata.authenticator, appID!!, false) }
                ?: emptyList()

        val allowedKeyIds = allowedMatchCriterias.flatMap { it.keyIDs ?: emptyList() }
        val disallowedKeyIds = disallowedMatchCriterias.flatMap { it.keyIDs ?: emptyList() }
        val filteredKeyIds = Crypto.getStoredKeyIds(appID!!, null)
                ?.filter { if (allowedKeyIds.isNotEmpty()) allowedKeyIds.contains(it) else true }
                ?.filter { !disallowedKeyIds.contains(it) }
                ?: emptyList()


        finalChallengeParams = FinalChallengeParams(appID!!, authenticationRequest.challenge, facetId, Util.moshi.adapter(ChannelBinding::class.java).fromJson(channelBinding)!!)

        if (filteredKeyIds.isEmpty()) {
            sendReturnIntent(UAFIntentType.UAF_OPERATION_RESULT, ErrorCode.NO_SUITABLE_AUTHENTICATOR, null)
            return
        }

        val authenticateIn = AuthenticateIn(
                appID = appID!!,
                finalChallenge = Base64.encodeToString(Util.moshi.adapter(FinalChallengeParams::class.java).toJson(finalChallengeParams).toByteArray(StandardCharsets.UTF_8), Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING),
                transaction = if (authenticationRequest.transaction != null && authenticationRequest.transaction.isNotEmpty()) {
                    authenticationRequest.transaction.find { it.contentType == "text/plain" }
                } else {
                    null
                },
                keyIDs = allowedKeyIds
        )

        this.sendToAsm(authenticateIn, authenticationRequest, sendToAsm)
    }

    private fun sendToAsm(authenticateIn: AuthenticateIn, authenticationRequest: UafAuthenticationRequest, sendToAsm: (String) -> Unit) {
        val asmRequestAuth = ASMRequestAuth(
                requestType = RequestType.Authenticate,
                asmVersion = authenticationRequest.header.upv,
                authenticatorIndex = 0,
                args = authenticateIn,
                exts = null
        )

        sendToAsm(Util.moshi.adapter(ASMRequestAuth::class.java).toJson(asmRequestAuth))
    }
}
